libraries/Resample/Resample.cpp

```diff
@@ -0,0 +1,205 @@
+    NE10_FREE(pState);
+    pCoeff = nullptr;
+    pState = nullptr;
+}
+
+int Decimator::setup(unsigned int decimationFactor, unsigned int blockSize, fir_quality quality, fir_phase phase) {
+    cleanup();
+    this->blockSizeIn = blockSize;
+    this->factor      = decimationFactor;
+    if (decimationFactor == 1) {
+        // Don't do anything
+        return 0;
+    }
+    std::vector<float> fir     = get_fir(decimationFactor, quality, phase);
+    uint               numTaps = fir.size();
+
+    pCoeff = (ne10_float32_t*)NE10_MALLOC(numTaps * sizeof(ne10_float32_t));
+    pState = (ne10_float32_t*)NE10_MALLOC(
+        (numTaps + blockSize - 1) * sizeof(ne10_float32_t));
+    if (!pState || !pCoeff)
+        return -1;
+
+    for (unsigned int n = 0; n < numTaps; ++n) {
+        // reverse filter coefficients
+        pCoeff[n] = fir[numTaps - 1 - n];
+    }
+    int err = ne10_fir_decimate_init_float(&decimator, numTaps, decimationFactor, pCoeff, pState, blockSize);
+    if (err) {
+        std::cerr << "ERR: blockSize must be whole multiple of decimationFactor" << std::endl;
+        return -1;
+    }
+
+    return 0;
+}
+
+void Decimator::process(ne10_float32_t* outBlock, const ne10_float32_t* inBlock) {
+    if (factor == 1 && outBlock != inBlock) {
+        memcpy(outBlock, inBlock, blockSizeIn * sizeof *outBlock);
+        return;
+    }
+    ne10_fir_decimate_float_neon(&decimator, (float*)inBlock, outBlock, blockSizeIn);
+}
+
+int Interpolator::setup(unsigned int L, unsigned int blockSize, fir_quality quality, fir_phase phase) {
+    cleanup();
+    this->blockSizeIn = blockSize / L;
+    this->factor      = L;
+    if (L == 1) {
+        // Don't do anything
+        return 0;
+    }
+    std::vector<float> fir      = get_fir(L, quality, phase);
+    uint               filtSize = fir.size();
+    // numTaps must be a multiple of the interpolation factor
+    uint numTaps     = ceil((float)filtSize / L) * L;
+    uint phaseLength = numTaps / L;
+    pCoeff           = (ne10_float32_t*)NE10_MALLOC(numTaps * sizeof(ne10_float32_t));
+    pState           = (ne10_float32_t*)NE10_MALLOC(
+        (blockSizeIn + phaseLength - 1) * sizeof(ne10_float32_t));
+    if (!pState || !pCoeff) {
+        return -1;
+    }
+
+    for (unsigned int n = 0; n < filtSize; ++n) {
+        // reverse fir coefficients
+        pCoeff[numTaps - filtSize + n] = fir[filtSize - 1 - n];
+    }
+    for (uint n = 0; n < numTaps - filtSize; ++n) {
+        // rest are are zeros
+        pCoeff[n] = 0;
+    }
+
+    int err = ne10_fir_interpolate_init_float(&interpolator, L, numTaps, pCoeff, pState, blockSizeIn);
+    if (err) {
+        std::cerr << "Error: couldn't init interpolator" << std::endl;
+        return -1;
+    }
+
+    return 0;
+}
+
+void Interpolator::process(ne10_float32_t* outBlock, const ne10_float32_t* inBlock) {
+    if (factor == 1 && outBlock != inBlock) {
+        memcpy(outBlock, inBlock, blockSizeIn * sizeof *outBlock);
```

Comment:
Could you use parenthesis for `sizeof`? I find it clearer that way especially in these non-trivial expressions.

Reply:
Yes.